thinkhazard_processing/processing.py

```diff
@@ -0,0 +1,145 @@
+
+import transaction
+import datetime
+import numpy
+import rasterio
+import pyproj
+from rasterio import features
+from shapely.ops import transform
+from functools import partial
+from geoalchemy2.shape import to_shape
+
+from thinkhazard_common.models import (
+    DBSession,
+    AdministrativeDivision,
+    HazardLevel,
+    )
+from .models import (
+    HazardSet,
+    Layer,
+    Output,
+    )
+from . import settings
+
+
+class ProcessException(Exception):
+    def __init__(self, message):
+        self.message = message
+
+
+def process_pendings():
+    ids = DBSession.query(HazardSet.id) \
+        .filter(HazardSet.complete.is_(True)) \
+        .filter(HazardSet.processed.is_(False))
+    for id in ids:
+        process_hazardset(id)
+
+
+def process_all():
+    ids = DBSession.query(HazardSet.id) \
+        .filter(HazardSet.complete.is_(True))
+    for id in ids:
+        process_hazardset(id, force=True)
+
+
+def process_hazardset(hazardset_id, force=False):
+    print hazardset_id
+    chrono = datetime.datetime.now()
+
+    hazardset = DBSession.query(HazardSet).get(hazardset_id)
+    if hazardset is None:
+        raise ProcessException('HazardSet {} does not exist.'
+                               .format(hazardset_id))
+
+    if hazardset.processed:
+        if force:
+            hazardset.processed = False
+            DBSession.flush()
+        else:
+            raise ProcessException('HazardSet {} has already been processed.'
+                                   .format(hazardset_id))
+
+    # lean previous outputs
+    DBSession.query(Output) \
+        .filter(hazardset_id == hazardset_id) \
+        .delete()
+
+    hazardtype = hazardset.hazardtype
+    hazardtype_settings = settings['hazard_types'][hazardtype.mnemonic]
+    threshold = hazardtype_settings['threshold']
+
+    project = partial(pyproj.transform,
+                      pyproj.Proj(init='epsg:3857'),
+                      pyproj.Proj(init='epsg:4326'))
+
+    print 'Reading raster data'
+    with rasterio.drivers():
+        # Register GDAL format drivers and configuration options with a
+        # context manager.
+        layers = []
+        for level in (u'HIG', u'MED', u'LOW'):
+            hazardlevel = HazardLevel.get(level)
+            layer = DBSession.query(Layer) \
+                .filter(Layer.hazardset_id == hazardset.id) \
+                .filter(Layer.hazardlevel_id == hazardlevel.id) \
+                .one()
+
+            with rasterio.open(layer.path()) as src:
+                src_data = src.read()
+                layer.shape = src.shape
+                layer.transform = src.transform
+            threshold_value = threshold[layer.hazardunit]
+            layer.data = (src_data > threshold_value).astype(rasterio.uint8)
+
+            layers.append(layer)
+
+    print 'Reading admin divisions'
+    admindivs = (
+        DBSession.query(AdministrativeDivision)
+        .filter(AdministrativeDivision.leveltype_id == 2)
```

Comment:
Admin levels are usually refered to as 0 for global,1 for state, 2 for regions but the leveltype_id sequence starts at 1 rather than 0 => shouldn't it be 3 in here ?